parquet-hadoop/src/main/java/org/apache/parquet/hadoop/DirectCodecPool.java

```diff
@@ -0,0 +1,239 @@
+      .synchronizedMap(new HashMap<Object, Object>());
+  private final Map<Class<?>, GenericObjectPool> cPools = (Map<Class<?>, GenericObjectPool>) (Object) Collections
+      .synchronizedMap(new HashMap<Object, Object>());
+
+  // Any of these can be null depending on the version of hadoop on the classpath
+  public static final Class DIRECT_DECOMPRESSION_CODEC_CLASS;
+  public static final Method DECOMPRESS_METHOD;
+  static final Method CREATE_DIRECT_DECOMPRESSOR_METHOD;
+
+  static {
+    Class tempClass = null;
+    Method tempCreateMethod = null;
+    Method tempDecompressMethod = null;
+    try {
+      tempClass = Class.forName("org.apache.hadoop.io.compress.DirectDecompressionCodec");
+      tempCreateMethod = tempClass.getMethod("createDirectDecompressor");
+      tempDecompressMethod = tempClass.getMethod("decompress", ByteBuffer.class, ByteBuffer.class);
+    } catch (ClassNotFoundException e) {
+      // do nothing, the class will just be assigned null
+    } catch (NoSuchMethodException e) {
+      // do nothing, the method will just be assigned null
+    }
+    DIRECT_DECOMPRESSION_CODEC_CLASS = tempClass;
+    CREATE_DIRECT_DECOMPRESSOR_METHOD = tempCreateMethod;
+    DECOMPRESS_METHOD = tempDecompressMethod;
+  }
+
+  private DirectCodecPool() {
+  }
+
+  public class CodecPool {
+    private final GenericObjectPool compressorPool;
+    private final GenericObjectPool decompressorPool;
+    private final GenericObjectPool directDecompressorPool;
+    private final boolean supportDirectDecompressor;
+
+    private CodecPool(final CompressionCodec codec){
+      try {
+        boolean supportDirectDecompressor = codec.getClass() == DIRECT_DECOMPRESSION_CODEC_CLASS;
+        compressorPool = new GenericObjectPool(new BasePoolableObjectFactory() {
+          public Object makeObject() throws Exception {
+            return codec.createCompressor();
+          }
+        }, Integer.MAX_VALUE);
+
+        Object com = compressorPool.borrowObject();
+        if (com != null) {
+          cPools.put(com.getClass(), compressorPool);
+          compressorPool.returnObject(com);
+        } else {
+          if (Log.WARN) {
+            LOG.warn("Unable to find compressor for codec " + codec.getClass().getName());
+          }
+        }
+
+        decompressorPool = new GenericObjectPool(new BasePoolableObjectFactory() {
+          public Object makeObject() throws Exception {
+            return codec.createDecompressor();
+          }
+        }, Integer.MAX_VALUE);
+
+        Object decom = decompressorPool.borrowObject();
+        if (decom != null) {
+          dePools.put(decom.getClass(), decompressorPool);
+          decompressorPool.returnObject(decom);
+        } else {
+          if (Log.WARN) {
+            LOG.warn("Unable to find decompressor for codec " + codec.getClass().getName());
+          }
+        }
+
+        if (supportDirectDecompressor) {
```

Comment:
we don't encapsulate this notion very well. There are several places where we check for this. This would go away once merge into the DirectCodecFactory